floodpipeline/data.py

```diff
@@ -0,0 +1,515 @@
+            "pcodes"
+        )  # pcodes of associated administrative divisions
+                
+        #self.basin_id: str = kwargs.get("basin_id") # associated basins/districts
+        
+class BasinDataUnit:
+    """Base class for basin data units"""
+
+    def __init__(self, **kwargs):
+        self.hybasid: str = kwargs.get("hybasid")
+        self.hybalevel: int = kwargs.get("hybalevel")
+        self.pcodes: Optional[dict] = kwargs.get("pcodes")
+        #self.lead_time: Optional[int] = kwargs.get("lead_time", None)
+
+class DischargeDataUnit(AdminDataUnit):
+    """River discharge data unit"""
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self.lead_time: int = kwargs.get("lead_time")
+        self.discharge_ensemble: List[float] = kwargs.get("discharge_ensemble", None)
+        self.discharge_mean: float = kwargs.get("discharge_mean", None)
+        self.pcode: str = kwargs.get("pcode", None)
+        if hasattr(self.discharge_ensemble, "__iter__"):
+            self.compute_mean()
+
+    def compute_mean(self):
+        """Compute mean river discharge"""
+        self.discharge_mean = sum(self.discharge_ensemble) / len(
+            self.discharge_ensemble
+        )
+
+
+class DischargeStationDataUnit(StationDataUnit):
+    """River discharge data unit - station"""
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self.lead_time: int = kwargs.get("lead_time")
+        self.discharge_ensemble: List[float] = kwargs.get("discharge_ensemble", None)
+        self.discharge_mean: float = kwargs.get("discharge_mean", None)
+        self.pcodes: dict = kwargs.get("pcodes", {})
+        self.lat: float = kwargs.get("lat", None)
+        self.lon: float = kwargs.get("lon", None)   
+        self.station_code: str = kwargs.get("station_code", None)
+        self.station_name: str = kwargs.get("station_name", None)
+        if hasattr(self.discharge_ensemble, "__iter__"):
+            self.compute_mean()
+
+    def compute_mean(self):
+        """Compute mean river discharge"""
+        self.discharge_mean = sum(self.discharge_ensemble) / len(
+            self.discharge_ensemble
+        )
+
+
+class RainfallBasinDataUnit(BasinDataUnit):
+    """Rainfall forecast data unit - basin"""
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self.lead_time: float = kwargs.get("lead_time")
+        self.rainfall_ensemble: List[float] = kwargs.get("rainfall_ensemble", [])
+        self.rainfall_mean: float = kwargs.get("rainfall_mean", None)
+        #self.hybasid: str = kwargs.get("hybasid")
+        #self.hybalevel: int = kwargs.get("hybalevel")
+        #self.hybalevel: int = self.select_hydrobasins_levle(self.lead_time)
+        if self.rainfall_ensemble:
+            self.compute_mean()
+
+    def compute_mean(self):
+        """Compute mean rainfall"""
+        self.rainfall_mean = sum(self.rainfall_ensemble) / len(self.rainfall_ensemble)
```

Comment:
This introduces the risk of a divide-by-zero error. [Here are a few ways to resolve the error](https://stackoverflow.com/a/29836987/1753041).